src/data-workspace/data-value-mutations/use-data-value-params.js

```diff
@@ -0,0 +1,33 @@
+import {
+    useContextSelection,
+    useApiAttributeParams,
+} from '../../shared/index.js'
+
+/**
+ * Formats a de/coc pair and the current context selection for use with the
+ * `api/dataValues` endpoint
+ */
+export const useDataValueParams = ({ deId, cocId }) => {
+    const [{ dataSetId, orgUnitId, periodId }] = useContextSelection()
+    const { attributeCombo, attributeOptions } = useApiAttributeParams()
+
+    const dataValueParams = {
+        de: deId,
+        co: cocId,
+        ds: dataSetId,
+        ou: orgUnitId,
+        pe: periodId,
+    }
+
+    // Add attribute params to context params if attribute is not default
+    // (useApiAttributeParams helper returns undefined props if attribute
+    // is default)
+    if (attributeCombo) {
+        // Get a ';'-separated listed of attribute options
+        const attributeOptionIdList = attributeOptions.join(';')
```

Comment:
a bit of an esoteric point here...but the way the attributeOptions information is being stored does not guarantee order (https://github.com/dhis2/data-entry-app/blob/a29c303083067162deb75917527f33d1c705fc7c/src/context-selection/attribute-option-combo-selector-bar-item/use-selected.js#L10), so in ART Monthly Summary, for example, if you click on "Implementing Partner" and then "Project" you'd get one order like `{LFsZ8v5v7rq: 'B3nxOazOO2G', yY2bQYqNt0o: 'i4Nbp8S2G6A'}`, then if you navigate to a different form, this gets cleared, and if you go back and then select "Project" first and then "Implementing Partner", you will get `{yY2bQYqNt0o: 'i4Nbp8S2G6A', LFsZ8v5v7rq: 'B3nxOazOO2G'}`...
so useApiAttributeParams just uses Object.values and then we join them, this can result in non-matching queryKeys (e.g. `{cp: 'i4Nbp8S2G6A;B3nxOazOO2G'}!={cp: 'B3nxOazOO2G;i4Nbp8S2G6A'}`. This then causes problems in the caching of these data values (notable e.g. in offline mode). Maybe we could alphabetize when generating queryKeys/mutationKeys?

Reply:
Astute observation! 😮 Alphabetizing them during query/mutation key formation would work, but what do you think about designing the attribute option selector to return the order of the `{ [category]: <attributeOptionId> }` object in the same order as the categories in the category combo? (e.g. always "Implementing Partner" before "Project")

Reply:
that sounds good! A bit more work to get them in that set order, but probably more understandable too if anyone looks at the api requests

Reply:
Made a Jira issue for this to address in another PR: https://dhis2.atlassian.net/browse/TECH-1342

Reply:
> designing the attribute option selector to return the order of the { [category]: <attributeOptionId> } object in the same order as the categories in the category combo? (e.g. always "Implementing Partner" before "Project")

If we were to do this, we would need to change the entire API, since it cannot be an object. JavaScript objects are not ordered - even if it kind of works to create an array, sort the properties and create the object - the order of object-properties are never guaranteed to be in order. 

So I think the easiest/best way is to just `.sort()` the `attributeOptionIdList` when it's used for the params.

Reply:
Chose to sort these here: https://github.com/dhis2/data-entry-app/blob/29740a6de1aa21deb89d4b3cdfb08bb1479d9dce/src/shared/use-api-attribute-params.js#L25-L26

Will close the Jira above